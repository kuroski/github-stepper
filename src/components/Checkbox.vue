<template>
  <div class="checkbox">
    <label :for="id" class="checkbox__content">
      <input
        :id="id"
        :value="value"
        :checked="value"
        type="checkbox"
        v-bind="$attrs"
        :aria-required="required"
        @change="$emit('input', $event.target.checked)"
      />
      <span>{{ label }}</span>
      <slot />
    </label>

    <ul v-if="errorMessages.length" class="checkbox__errors">
      <li v-for="error in errorMessages" :key="error">{{ error }}</li>
    </ul>
  </div>
</template>

<script>
export default {
  name: "TextField",
  props: {
    id: {
      type: String,
      default: undefined
    },
    value: {
      type: Boolean,
      required: true
    },
    label: {
      type: String,
      default: ""
    },
    required: {
      type: Boolean,
      default: false
    },
    errorMessages: {
      type: Array,
      default: () => []
    }
  }
};
</script>

<style scoped>
.checkbox + .checkbox {
  margin-top: var(--space-5);
}

.checkbox {
  display: flex;
  flex-direction: column;
}

.checkbox__content {
  position: relative;
  display: flex;
  align-items: center;
}

.checkbox__content > input {
  margin-right: var(--space-2);
  padding: var(--space-2) var(--space-3);
  border: solid 1px var(--color-gray-500);
  color: var(--color-blue-900);
}

.checkbox__content > input::placeholder {
  color: var(--color-gray-500);
}

.checkbox__errors {
  margin-top: var(--space-2);
  margin-bottom: 0;
  list-style-position: inside;
  padding-left: 0;
  color: var(--color-red-600);
}

.checkbox__errors > li:not(:first-child) {
  margin-top: var(--space-1);
}

.checkbox__progress {
  position: relative;
  height: 2px;
  overflow: hidden;
  background-color: var(--color-blue-200);
}

.checkbox__progress::before {
  content: "";
  position: absolute;
  width: 100%;
  height: 100%;
  background-color: var(--color-blue-600);
  animation: indeterminate 1s infinite linear;
  transform-origin: 0% 50%;
}

@keyframes indeterminate {
  0% {
    transform: translateX(0) scaleX(0);
  }
  40% {
    transform: translateX(0) scaleX(0.4);
  }
  100% {
    transform: translateX(100%) scaleX(0.5);
  }
}
</style>
