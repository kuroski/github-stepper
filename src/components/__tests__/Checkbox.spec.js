import { renderWithDependencies } from "@unit/helpers";
import { fireEvent } from "@testing-library/vue";
import Checkbox from "@/components/Checkbox";

describe("Checkbox", () => {
  it("renders a regular input", async () => {
    const { html, getByLabelText, emitted } = renderWithDependencies(Checkbox, {
      props: {
        id: "terms",
        value: false,
        label: "Do you aceept our terms?"
      }
    });

    const checkboxInput = getByLabelText("Do you aceept our terms?");

    await fireEvent.click(checkboxInput);

    expect(html()).toMatchSnapshot();
    expect(checkboxInput).toBeInTheDocument();
    expect(checkboxInput).toHaveAttribute("id", "terms");
    expect(emitted().input).toHaveLength(1);
    expect(emitted().input[0][0]).toEqual(true);
  });

  it("displays error messages", () => {
    const { getByText } = renderWithDependencies(Checkbox, {
      props: {
        id: "terms",
        value: "",
        label: "Do you aceept our terms?",
        errorMessages: ["This field is required"]
      }
    });

    expect(getByText("This field is required")).toBeInTheDocument();
  });
});
