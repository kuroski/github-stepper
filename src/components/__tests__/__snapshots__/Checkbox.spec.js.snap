// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders a regular input 1`] = `
<div class="checkbox"><label for="terms" class="checkbox__content"><input id="terms" type="checkbox" value="false"> <span>Do you aceept our terms?</span> </label>
  <!---->
</div>
`;
